### ISSUE TITLE
unable to service request at this time

### ISSUE THREAD

**Issue by j0110** [mullvad/wgephemeralpeer#5]
Hi,
When using this executable to upgrade the tunnel connection, I often get this error:
unable to connect ephemeral peer, rpc error: code = Unknown desc = unable to service request at this time
If I wait a minute, it works again, but I can't understand why it does nor work the first time.
Thanks

**Comment by mvd-ows** [comment]
This is currently being looked into. Will update here once resolved.

**Comment by mvd-ows** [comment]
We've corrected a config issue in the general area where your client request is being processed. However, the specific error message you're seeing should not be produced even running with the non-optimal config.

Would you mind sharing some examples of relays where you've seen this error?

**Comment by j0110** [comment]
Tested today, and I got the same issue, it is with fr-par-wg-001. But I got the same issue with other servers too.

For info, I use this script to execute the upgrade : [https://github.com/j0110/mullvad-controller/blob/main/starter.py](https://github.com/j0110/mullvad-controller/blob/main/starter.py)

**Comment by mvd-ows** [comment]
Trying to upgrade an already upgraded peer will be rejected by the server. This is likely what's happening here.

The error is being returned to the client with a generic error message to prevent exploration of server state.

You could extend your script to check if the active peer already has a PSK configured and then avoid attempting a second update.

**Comment by j0110** [comment]
Well, this script is used as as Post-Activation so no PSK is set at that time...
If the script is run as Post-Activation it fails, but if I wait few seconds and run it manually, it works... Strange !

**Comment by j0110** [comment]
I got surprisingly a new error :
unable to connect to relay, ensure you are able to connect to 10.64.0.1 on TCP port 1337
Strange, because 10.64.0.1 is pingable before trying to attempt to upgrade the tunnel... Same error even if I had a delay of 1sec just after the ping test is successful.

**Comment by mvd-ows** [comment]
You local firewall is likely interfering with communications. Can you confirm that `mullvad-upgrade-tunnel` does indeed work when used with `wg-quick`, without all the extra steps?

**Comment by mvd-ows** [comment]
Alright, you seem to have a firewall and/or routing issue. Remove the PostUp and with a tunnel established, ensure that you can connect to 10.64.0.1:1337 over TCP. Since you're on Windows, use curl or ncat to test connectivity.

**Comment by j0110** [comment]
Well, 10.64.0.1:1337 is accessible.
I can perform manually the tunnel upgrade after the tunnel was established, but it won't works as a PostActivation action.
Therefore I don't think it is a firewall/routing issue.

**Comment by mvd-ows** [comment]
Interesting. We'll do some testing on this end and get back to you.

**Comment by mvd-ows** [comment]
We've retested this on Win10 and Win11 and found no issues.

Keep in mind that the PostUp action runs in the context of the SYSTEM user. Is it possible that you could have configuration or security software that prevents that user from making the required connection?

In the event that there's a timing issue involved, it may help to insert a pause before attempting to upgrade the tunnel.

There are some restrictions related to this, e.g. it's not trivial to chain PostUp commands using the Windows implementation of Wireguard, and also, the `timeout` system utility cannot be used in this context.

Create a separate script file `hack-upgrade-tunnel.bat` with the following content:

```
powershell -nop -c "& {sleep 3}"
mullvad-upgrade-tunnel -wg-interface %1
```

And you'll need to adjust the PostUp:

`PostUp = hack-upgrade-tunnel.bat %WIREGUARD_TUNNEL_NAME%`

Let us know if that makes any difference.

**Comment by j0110** [comment]
I didn't got the same error
`2024-10-08 15:16:38.172: [TUN] [m_FR.par] cmd> unable to connect to relay, ensure you are able to connect to 10.64.0.1 on TCP port 1337`
I pinged 10.64.0.1 during the connection process and it is accessible.

I want just to say again that this issue appeared few weeks ago. Maybe something changed on servers because I used to connect with this exact configuration before.

**Comment by mvd-ows** [comment]
You say that you're able to manually perform the tunnel upgrade once the tunnel is established. Can you try doing this in the context of the SYSTEM user?

You can use PsExec to spawn a shell in the correct context. Get the tool from here: https://learn.microsoft.com/en-us/sysinternals/downloads/psexec

This should be the command line to use:

`psexec64 -i -s cmd.exe`

**Comment by j0110** [comment]
It works with :
`PostUp = ""C:\Path\to\psexec64.exe" -i -s -accepteula -nobanner "C:\Path\to\mullvad-controller\mullvad-upgrade-tunnel.exe" -wg-interface %WIREGUARD_TUNNEL_NAME%"`
May be a change could be made to mull-upgrade-tunnel to include this thing ?

**Comment by mvd-ows** [comment]
That doesn't really make much sense. WireGuard should already be running in that context which you are now forcing.

The only difference I'm aware of, would be this functionality, which has been activated before running PostUp: https://github.com/WireGuard/wireguard-windows/blob/e70799b1440690e7d4140bffc7c73baf903c7b54/elevate/privileges.go#L16

Nevertheless, the regular PostUp configuration seems to work for everyone except for you. Is there a possibility that you could have security/networking software installed that could interfere?

**Comment by j0110** [comment]
I double-checked but I don't have any security softwares that might block something.
It is really surprising because if I put `whoami` in the `PostUp` line, I got `NT Authority\SYSTEM`.
Is there a possibility that `mullvad-upgrade-tunnel` script loose its privileges at some time ?

**Comment by mvd-ows** [comment]
The only thing I can think of are the process token adjustments I linked previously. This is likely inherited by the PostUp-launched process.

Have you been able to reproduce this issue on more than a single machine? As I mentioned, we're unable to reproduce it on our end. What's the exact version of Windows you're using?

Did you try the script I provided that merely introduces a delay? It seems to me that your current solution similarly introduces a tiny delay and restores a few irrelevant token privileges.
